src/Framework/MockObject/Runtime/Api/DoubledCloneMethod.php

```diff
@@ -16,6 +16,6 @@ trait DoubledCloneMethod
 {
     public function __clone(): void
     {
-        $this->__phpunit_invocationMocker = clone $this->__phpunit_getInvocationHandler();
+        $this->__phpunit_stubInternalState->cloneInvocationHandler();
```

Comment:
Unfortunately, I identified a bug in the implementation, for a behavior that tests are not currently covering. I have added the tests now in #5665 that would fail with current code.
The correct code with using a readonly property for the internal state should be:
```suggestion
        $this->__phpunit_stubInternalState = clone $this->__phpunit_stubInternalState;
        $this->__phpunit_stubInternalState->cloneInvocationHandler();
```  
  
  
Cloning readonly properties does not work in PHP 8.1 and PHP 8.2, so this kind of implementation can only exist in a PHPUnit version that has a minim PHP supported version as PHP 8.3.

I think we can avoid the limitation by changing the internal state implementation to use a WeakMap for the invocation handler, with test doubles as keys, so we can properly clone it.

Reply:
I applied my suggested change to have the code without bugs and fix at least PHP 8.3.
If anyone finds a solution, feel free to suggest something.

Reply:
We should not try to `clone` the test double object when `readonly` properties are involved. Instead, before the `clone` statement is reached, we should raise a PHPUnit error with an explanation that cloning test double objects that have `readonly` properties requires PHP >= 8.3.

Reply:
Yes, thank you, the idea is fine and blocks only cloning when the mock is for a readonly class.
But I think this also implies that we should create readonly properties for internal state only when mocking readonly classes? With current implementation, we have the internal state as a readonly property for all cases.
I'm thinking that having a separate Trait for readonly class vs not readonly classes would not be nice, even if keeping differences at minimum only for the property definition. Do you have an idea for this?